=== modified file 'Percona-Server/client/mysqladmin.cc'
--- Percona-Server/client/mysqladmin.cc	2013-02-12 07:47:19 +0000
+++ Percona-Server/client/mysqladmin.cc	2013-04-15 07:08:10 +0000
@@ -23,6 +23,7 @@
 #include <my_pthread.h>				/* because of signal()	*/
 #include <sys/stat.h>
 #include <mysql.h>
+#include <password.h>       /* my_make_scrambled_password_323, my_make_scrambled_password */
 #include <sql_common.h>
 #include <welcome_copyright_notice.h>           /* ORACLE_WELCOME_COPYRIGHT_NOTICE */
 #include <mysqld_error.h>                       /* to check server error codes */
@@ -1045,9 +1046,9 @@
           we will give one more try with old format.
         */
         if (old)
-          make_scrambled_password_323(crypted_pw, typed_password);
+          my_make_scrambled_password_323(crypted_pw, typed_password, strlen(typed_password));
         else
-          make_scrambled_password(crypted_pw, typed_password);
+          my_make_scrambled_password_sha1(crypted_pw, typed_password, strlen(typed_password));
       }
       else
 	crypted_pw[0]=0;			/* No password */

=== modified file 'Percona-Server/include/my_sys.h'
--- Percona-Server/include/my_sys.h	2013-03-05 12:46:43 +0000
+++ Percona-Server/include/my_sys.h	2013-04-15 07:08:10 +0000
@@ -261,6 +261,7 @@
 
 /* charsets */
 #define MY_ALL_CHARSETS_SIZE 2048
+#define default_charset_info mysql_default_charset_info  /* namespace sanity */
 extern MYSQL_PLUGIN_IMPORT CHARSET_INFO *default_charset_info;
 extern MYSQL_PLUGIN_IMPORT CHARSET_INFO *all_charsets[MY_ALL_CHARSETS_SIZE];
 extern CHARSET_INFO compiled_charsets[];

=== modified file 'Percona-Server/include/mysql.h.pp'
--- Percona-Server/include/mysql.h.pp	2013-02-12 07:47:19 +0000
+++ Percona-Server/include/mysql.h.pp	2013-04-15 07:08:10 +0000
@@ -90,7 +90,7 @@
 void my_net_local_init(NET *net);
 void net_end(NET *net);
 void net_clear(NET *net, my_bool check_buffer);
-my_bool net_realloc(NET *net, size_t length);
+my_bool mysql_net_realloc(NET *net, size_t length);
 my_bool net_flush(NET *net);
 my_bool my_net_write(NET *net,const unsigned char *packet, size_t len);
 my_bool net_write_command(NET *net,unsigned char command,
@@ -129,13 +129,11 @@
 double my_rnd(struct rand_struct *);
 void create_random_string(char *to, unsigned int length, struct rand_struct *rand_st);
 void hash_password(unsigned long *to, const char *password, unsigned int password_len);
-void make_scrambled_password_323(char *to, const char *password);
 void scramble_323(char *to, const char *message, const char *password);
 my_bool check_scramble_323(const unsigned char *reply, const char *message,
                            unsigned long *salt);
 void get_salt_from_password_323(unsigned long *res, const char *password);
 void make_password_from_salt_323(char *to, const unsigned long *salt);
-void make_scrambled_password(char *to, const char *password);
 void scramble(char *to, const char *message, const char *password);
 my_bool check_scramble(const unsigned char *reply, const char *message,
                        const unsigned char *hash_stage2);

=== modified file 'Percona-Server/include/mysql_com.h'
--- Percona-Server/include/mysql_com.h	2013-03-05 12:46:43 +0000
+++ Percona-Server/include/mysql_com.h	2013-04-15 07:08:10 +0000
@@ -497,6 +497,7 @@
 void my_net_local_init(NET *net);
 void net_end(NET *net);
 void net_clear(NET *net, my_bool check_buffer);
+#define net_realloc mysql_net_realloc    /* namespace sanity */
 my_bool net_realloc(NET *net, size_t length);
 my_bool	net_flush(NET *net);
 my_bool	my_net_write(NET *net,const unsigned char *packet, size_t len);
@@ -574,14 +575,12 @@
 void create_random_string(char *to, unsigned int length, struct rand_struct *rand_st);
 
 void hash_password(unsigned long *to, const char *password, unsigned int password_len);
-void make_scrambled_password_323(char *to, const char *password);
 void scramble_323(char *to, const char *message, const char *password);
 my_bool check_scramble_323(const unsigned char *reply, const char *message,
                            unsigned long *salt);
 void get_salt_from_password_323(unsigned long *res, const char *password);
 void make_password_from_salt_323(char *to, const unsigned long *salt);
 
-void make_scrambled_password(char *to, const char *password);
 void scramble(char *to, const char *message, const char *password);
 my_bool check_scramble(const unsigned char *reply, const char *message,
                        const unsigned char *hash_stage2);

=== modified file 'Percona-Server/libmysql/CMakeLists.txt'
--- Percona-Server/libmysql/CMakeLists.txt	2013-04-15 06:52:16 +0000
+++ Percona-Server/libmysql/CMakeLists.txt	2013-04-15 07:08:10 +0000
@@ -224,7 +224,7 @@
       SET(libmysql_link_flags
         "${libmysql_link_flags} ${LINK_FLAG_NO_UNDEFINED}")
       SET(libmysql_link_flags
-        "${libmysql_link_flags} -Wl,--version-script=libmysql.ver")
+        "${libmysql_link_flags} -Wl,--version-script=${CMAKE_CURRENT_SOURCE_DIR}/libmysql.map")
       SET_TARGET_PROPERTIES(libmysql
         PROPERTIES LINK_FLAGS "${libmysql_link_flags}")
     ENDIF() 

