--- /dev/null
+++ b/mysql-test/r/percona_innodb_version.result
@@ -0,0 +1,3 @@
+SELECT @@innodb_version;
+@@innodb_version
+1.1.x-x.x
--- /dev/null
+++ b/mysql-test/t/percona_innodb_version.test
@@ -0,0 +1,8 @@
+--source include/have_innodb.inc
+
+########################################################################
+# Bug #917246: Unintentional change of innodb_version format in 5.1.60
+########################################################################
+
+--replace_regex /1\.1\.[0-9]+-[0-9]+\.[0-9]+/1.1.x-x.x/
+SELECT @@innodb_version;
