--- a/sql/debug_sync.cc
+++ b/sql/debug_sync.cc
@@ -1737,7 +1737,7 @@
     if (action->wait_for.length())
     {
       mysql_mutex_t *old_mutex;
-      mysql_cond_t  *old_cond;
+      mysql_cond_t*  UNINIT_VAR(old_cond);
       int             error= 0;
       struct timespec abstime;
 
--- a/storage/innobase/ibuf/ibuf0ibuf.c
+++ b/storage/innobase/ibuf/ibuf0ibuf.c
@@ -1523,7 +1523,7 @@
 
 	for (i = 0; i < IBUF_OP_COUNT; i++) {
 #ifdef HAVE_ATOMIC_BUILTINS
-		os_atomic_increment_ulint(&arr[i], ops[i]);
+		(void) os_atomic_increment_ulint(&arr[i], ops[i]);
 #else /* HAVE_ATOMIC_BUILTINS */
 		arr[i] += ops[i];
 #endif /* HAVE_ATOMIC_BUILTINS */
@@ -4785,7 +4785,7 @@
 	mem_heap_free(heap);
 
 #ifdef HAVE_ATOMIC_BUILTINS
-	os_atomic_increment_ulint(&ibuf->n_merges, 1);
+	(void) os_atomic_increment_ulint(&ibuf->n_merges, 1);
 	ibuf_add_ops(ibuf->n_merged_ops, mops);
 	ibuf_add_ops(ibuf->n_discarded_ops, dops);
 #else /* HAVE_ATOMIC_BUILTINS */
