--- a/sql/log.cc
+++ b/sql/log.cc
@@ -1905,7 +1905,7 @@
 
 static int find_uniq_filename(char *name)
 {
-  long                  number;
+  long                  UNINIT_VAR(number);
   uint                  i;
   char                  buff[FN_REFLEN];
   struct st_my_dir     *dir_info;
--- a/sql/sql_table.cc
+++ b/sql/sql_table.cc
@@ -7140,7 +7140,9 @@
     need_copy_table= ALTER_TABLE_DATA_CHANGED;
   else
   {
-    enum_alter_table_change_level need_copy_table_res;
+    enum_alter_table_change_level need_copy_table_res=
+        ALTER_TABLE_METADATA_ONLY;
+
     /* Check how much the tables differ. */
     if (compare_tables(table, alter_info,
                        create_info, order_num,
