diff -ruN a/mysql-test/r/plugin_auth.result b/mysql-test/r/plugin_auth.result
--- a/mysql-test/r/plugin_auth.result	2010-11-03 03:01:12.000000000 +0500
+++ b/mysql-test/r/plugin_auth.result	2010-11-16 10:56:04.000000000 +0500
@@ -14,9 +14,9 @@
 test proxies_priv columns
 SELECT * FROM mysql.proxies_priv;
 Host	User	Proxied_host	Proxied_user	With_grant	Grantor	Timestamp
-localhost	root			1		xx
-unknown	root			1		xx
-%	plug	%	plug_dest	0	root@localhost	xx
+host	root			1		xx
+host	root			1		xx
+host	plug	%	plug_dest	0	root@localhost	xx
 test mysql.proxies_priv;
 SHOW CREATE TABLE mysql.proxies_priv;
 Table	Create Table
diff -ruN a/mysql-test/t/plugin_auth.test b/mysql-test/t/plugin_auth.test
--- a/mysql-test/t/plugin_auth.test	2010-11-03 03:01:12.000000000 +0500
+++ b/mysql-test/t/plugin_auth.test	2010-11-16 10:56:22.000000000 +0500
@@ -17,7 +17,7 @@
 
 GRANT PROXY ON plug_dest TO plug;
 --echo test proxies_priv columns
---replace_column 7 xx
+--replace_column 1 host 7 xx;
 SELECT * FROM mysql.proxies_priv;
 --echo test mysql.proxies_priv;
 SHOW CREATE TABLE mysql.proxies_priv;
