diff -Naur a/HandlerSocket-Plugin-for-MySQL/configure.ac b/HandlerSocket-Plugin-for-MySQL/configure.ac
--- a/HandlerSocket-Plugin-for-MySQL/configure.ac	2011-04-12 06:01:24.101637007 -0300
+++ b/HandlerSocket-Plugin-for-MySQL/configure.ac	2011-04-12 07:08:56.761637002 -0300
@@ -101,14 +101,10 @@
   AC_ARG_WITH([mysql-plugindir],
     [AS_HELP_STRING([--with-mysql-plugindir=PATH], [MySQL plugin directory where handlersocket.so to be copied])],
     [
-      ac_mysql_plugin_dir=`cd $withval && pwd`
-      if test -d "$ac_mysql_plugin_dir/" ; then
-        PLUGIN_DIR="$ac_mysql_plugin_dir"
-        AC_SUBST(PLUGIN_DIR)
-        AC_MSG_RESULT([yes: Using $ac_mysql_plugin_dir])
-      else
-        AC_MSG_ERROR([invalid MySQL plugin directory : $ac_mysql_plugin_dir])
-      fi
+      ac_mysql_plugin_dir=`echo $withval`
+      PLUGIN_DIR="$ac_mysql_plugin_dir"
+      AC_SUBST(PLUGIN_DIR)
+      AC_MSG_RESULT([yes: Using $ac_mysql_plugin_dir])
     ],
     [
       LIB_DIR_TMP=`"$ac_mysql_config" --plugindir`
