diff -ruN a/mysql-test/mysql-test-run.pl b/mysql-test/mysql-test-run.pl
--- a/mysql-test/mysql-test-run.pl	2011-04-12 15:11:40.000000000 +0300
+++ b/mysql-test/mysql-test-run.pl	2011-07-06 14:23:15.367219807 +0300
@@ -157,7 +157,7 @@
 # executables will be used by the test suite.
 our $opt_vs_config = $ENV{'MTR_VS_CONFIG'};
 
-my $DEFAULT_SUITES= "main,binlog,federated,rpl,rpl_ndb,ndb,innodb,innodb_plugin";
+my $DEFAULT_SUITES= "main,binlog,federated,rpl,rpl_ndb,ndb,innodb,innodb_plugin,percona";
 my $opt_suites;
 
 our $opt_verbose= 0;  # Verbose output, enable with --verbose
