diff -ruN a/mysql-test/mysql-test-run.pl b/mysql-test/mysql-test-run.pl
--- a/mysql-test/mysql-test-run.pl	2011-05-11 14:54:11.000000000 +0300
+++ b/mysql-test/mysql-test-run.pl	2011-07-06 14:27:56.931566255 +0300
@@ -162,7 +162,7 @@
 
 # If you add a new suite, please check TEST_DIRS in Makefile.am.
 #
-my $DEFAULT_SUITES= "main,sys_vars,binlog,federated,rpl,innodb,perfschema";
+my $DEFAULT_SUITES= "main,sys_vars,binlog,federated,rpl,innodb,perfschema,percona";
 my $opt_suites;
 
 our $opt_verbose= 0;  # Verbose output, enable with --verbose
