--- client/CMakeLists.txt	2018-08-18 03:29:59.708213642 -0400
+++ client/CMakeLists.txt	2018-08-18 03:29:59.712547041 -0400
@@ -40,7 +40,7 @@
   multi_option.cc
   ${CMAKE_CURRENT_SOURCE_DIR}/common/user_registration.cc
   ${CMAKE_SOURCE_DIR}/sql-common/sql_string.cc
-  LINK_LIBRARIES perconaserverclient ${EDITLINE_LIBRARY}
+  LINK_LIBRARIES mysqlclient ${EDITLINE_LIBRARY}
   )

 MYSQL_ADD_EXECUTABLE(mysqltest
@@ -57,7 +57,7 @@
   COMPONENT Test
   DEPENDENCIES GenError GenClientError
   ENABLE_EXPORTS
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )
 IF (WITH_COREDUMPER)
   TARGET_LINK_LIBRARIES(mysqltest coredumper)
@@ -67,23 +67,23 @@
   check/mysqlcheck.cc
   check/mysqlcheck_core.cc
   multi_factor_passwordopt-vars.cc
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )
 MYSQL_ADD_EXECUTABLE(mysqldump
   mysqldump.cc
   multi_factor_passwordopt-vars.cc
   multi_option.cc
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )
 MYSQL_ADD_EXECUTABLE(mysqlimport
   mysqlimport.cc
   multi_factor_passwordopt-vars.cc
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )
 MYSQL_ADD_EXECUTABLE(mysqlshow
   mysqlshow.cc
   multi_factor_passwordopt-vars.cc
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )

 OPTION(WITH_JSON_BINLOG_LIBRARY
@@ -240,7 +240,7 @@

 SET(MYSQLBINLOG_LIBRARIES
   mysql_binlog_event
-  perconaserverclient
+  mysqlclient
   )

 IF(WITH_JSON_BINLOG_LIBRARY)
@@ -282,20 +282,20 @@
 MYSQL_ADD_EXECUTABLE(mysqladmin
   mysqladmin.cc
   multi_factor_passwordopt-vars.cc
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )
 MYSQL_ADD_EXECUTABLE(mysqlslap
   mysqlslap.cc
   multi_factor_passwordopt-vars.cc
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )
 MYSQL_ADD_EXECUTABLE(mysql_config_editor
   mysql_config_editor.cc
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )
 MYSQL_ADD_EXECUTABLE(mysql_secure_installation
   mysql_secure_installation.cc
-  LINK_LIBRARIES perconaserverclient
+  LINK_LIBRARIES mysqlclient
   )

 # "WIN32" also covers 64 bit. "echo" is used in some files below "mysql-test/".
--- client/migrate_keyring/CMakeLists.txt
+++ client/migrate_keyring/CMakeLists.txt
@@ -41,7 +41,7 @@
   ${CMAKE_DL_LIBS}
   minchassis
   mysys
-  perconaserverclient
+  mysqlclient
   OpenSSL::SSL OpenSSL::Crypto
 )

--- include/mysql.h	2022-09-13 10:19:20.000000000 +0300
+++ include/mysql.h	2022-09-13 12:40:29.000000000 +0300
@@ -28,7 +28,7 @@
 /**
   @file include/mysql.h
   This file defines the client API to MySQL and also the ABI of the
-  dynamically linked libperconaserverclient.
+  dynamically linked libmysqlclient.

   The ABI should never be changed in a released product of MySQL,
   thus you need to take great care when changing the file. In case
@@ -410,12 +410,12 @@

 /*
   mysql_server_init/end need to be called when using libmysqld or
-  libperconaserverclient (exactly, mysql_server_init() is called by
+  libmysqlclient (exactly, mysql_server_init() is called by
   mysql_init() so you don't need to call it explicitly; but you need to call
   mysql_server_end() to free memory). The names are a bit misleading
   (mysql_SERVER* to be used when using libmysqlCLIENT). So we add more general
   names which suit well whether you're using libmysqld or
-  libperconaserverclient. We intend to promote these aliases over the
+  libmysqlclient. We intend to promote these aliases over the
   mysql_server* ones.
 */
 #define mysql_library_init mysql_server_init
--- libmysql/CMakeLists.txt
+++ libmysql/CMakeLists.txt
@@ -304,15 +304,15 @@
 # authentication mysql_native_password client plug-in
 ADD_SUBDIRECTORY(authentication_native_password)

-# Merge several convenience libraries into one big perconaserverclient
-MERGE_CONVENIENCE_LIBRARIES(perconaserverclient ${LIBS_TO_MERGE}
+# Merge several convenience libraries into one big mysqlclient
+MERGE_CONVENIENCE_LIBRARIES(mysqlclient ${LIBS_TO_MERGE}
   COMPONENT Development
   LINK_LIBRARIES ${LIBS_TO_LINK}
   )

 # Visual Studio users need debug  static library for debug projects
 IF(MSVC)
-  INSTALL_DEBUG_TARGET(perconaserverclient DESTINATION ${INSTALL_LIBDIR}/debug)
+  INSTALL_DEBUG_TARGET(mysqlclient DESTINATION ${INSTALL_LIBDIR}/debug)

   INSTALL_DEBUG_TARGET(auth_win_client DESTINATION ${INSTALL_LIBDIR}/debug)
   INSTALL_DEBUG_TARGET(clientlib DESTINATION ${INSTALL_LIBDIR}/debug)
@@ -326,13 +326,13 @@
     STRING_APPEND(OS_SHARED_LIB_VERSION
       ".${SHARED_LIB_MINOR_VERSION}.${SHARED_LIB_PATCH_VERSION}")
   ENDIF()
-  # Name of shared library is libperconaserverclient on Unix
-  SET(UNIX_OUTPUT_NAME OUTPUT_NAME perconaserverclient)
+  # Name of shared library is libmysqlclient on Unix
+  SET(UNIX_OUTPUT_NAME OUTPUT_NAME mysqlclient)
   SET(UNIX_VERSION     VERSION     "${OS_SHARED_LIB_VERSION}")
   SET(UNIX_SOVERSION   SOVERSION   "${SHARED_LIB_MAJOR_VERSION}")
 ENDIF()
 
-# Merge several convenience libraries into one big perconaserverclient
+# Merge several convenience libraries into one big mysqlclient
 # and link them together into shared library.
 MERGE_LIBRARIES_SHARED(libmysql ${LIBS_TO_MERGE}
   EXPORTS
@@ -353,7 +353,7 @@
 
 IF(LINUX_STANDALONE AND KERBEROS_CUSTOM_LIBRARY)
   ADD_DEPENDENCIES(libmysql ${kerberos_target})
-  ADD_DEPENDENCIES(perconaserverclient ${kerberos_target})
+  ADD_DEPENDENCIES(mysqlclient ${kerberos_target})
 ENDIF()
 
 IF(UNIX)
@@ -368,8 +368,8 @@
   ENDIF()
 
   # clean direct output needs to be set several targets have the same name
-  #(perconaserverclient in this case)
-  SET_TARGET_PROPERTIES(perconaserverclient PROPERTIES CLEAN_DIRECT_OUTPUT 1)
+  #(mysqlclient in this case)
+  SET_TARGET_PROPERTIES(mysqlclient PROPERTIES CLEAN_DIRECT_OUTPUT 1)
   SET_TARGET_PROPERTIES(libmysql PROPERTIES CLEAN_DIRECT_OUTPUT 1)
 ENDIF()
 
--- libmysql/libmysql.ver.in
+++ libmysql/libmysql.ver.in
@@ -21,5 +21,5 @@
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
 
 
-libperconaserverclient_@SHARED_LIB_MAJOR_VERSION@.0
+libmysqlclient_@SHARED_LIB_MAJOR_VERSION@.0
 { global: ${CLIENT_API_FUNCTIONS};${CLIENT_API_FUNCTIONS_UNDOCUMENTED}; ${CLIENT_API_NONBLOCKING_FUNCTIONS}; local: *; };
--- plugin/percona-pam-for-mysql/CMakeLists.txt
+++ plugin/percona-pam-for-mysql/CMakeLists.txt
@@ -36,7 +36,7 @@ IF(HAVE_PAM AND HAVE_GETPWNAM_R AND HAVE_GETGRGID_R AND HAVE_DLFCN_H)
   MYSQL_ADD_PLUGIN(dialog
                    src/dialog.cc
                    ../../sql-common/get_password.cc
-                   LINK_LIBRARIES perconaserverclient
+                   LINK_LIBRARIES mysqlclient
                    MODULE_ONLY)
 
   IF(UNIX)
--- plugin/x/tests/driver/CMakeLists.txt
+++ plugin/x/tests/driver/CMakeLists.txt
@@ -41,7 +41,7 @@
   LINK_LIBRARIES
   mysys
   strings
-  perconaserverclient
+  mysqlclient
   ${MYSQLX_CLIENT_FULL_LIB}
   ${MYSQLX_PROTOCOL_FULL_LIB}
   ext::libprotobuf
--- router/src/router/src/CMakeLists.txt
+++ router/src/router/src/CMakeLists.txt
@@ -53,7 +53,7 @@
   PUBLIC
   harness_stdx
   harness-library  # log_debug
-  perconaserverclient
+  mysqlclient
   )

 ROUTER_ADD_SHARED_LIBRARY(router_cluster
--- scripts/CMakeLists.txt
+++ scripts/CMakeLists.txt
@@ -302,7 +302,7 @@ ELSE()
 ENDIF()
 
 # Use cmake variables to inspect dependencies for
-# perconaserverclient library (add -l stuff)
+# mysqlclient library (add -l stuff)
 SET(CLIENT_LIBS "")
 SET(LIBS "")
 
@@ -337,11 +337,11 @@ MACRO(EXTRACT_LINK_LIBRARIES target var)
   STRING(REGEX REPLACE "[ ]+$" "" ${var} "${${var}}")
 ENDMACRO()
 
-EXTRACT_LINK_LIBRARIES(perconaserverclient CLIENT_LIBS)
+EXTRACT_LINK_LIBRARIES(mysqlclient CLIENT_LIBS)
 
 IF(MSVC)
-  GET_TARGET_PROPERTY(LIBMYSQL_OS_SHLIB_VERSION perconaserverclient VERSION)
-  GET_TARGET_PROPERTY(LIBMYSQL_OS_OUTPUT_NAME perconaserverclient OUTPUT_NAME)
+  GET_TARGET_PROPERTY(LIBMYSQL_OS_SHLIB_VERSION mysqlclient VERSION)
+  GET_TARGET_PROPERTY(LIBMYSQL_OS_OUTPUT_NAME mysqlclient OUTPUT_NAME)
 ELSE()
   GET_TARGET_PROPERTY(LIBMYSQL_OS_SHLIB_VERSION libmysql VERSION)
   GET_TARGET_PROPERTY(LIBMYSQL_OS_OUTPUT_NAME libmysql OUTPUT_NAME)
--- storage/ndb/test/run-test/CMakeLists.txt
+++ storage/ndb/test/run-test/CMakeLists.txt
@@ -47,7 +47,7 @@
   test_execution_resources.cpp
   DESTINATION mysql-test/ndb
   ENABLE_EXPORTS
-  LINK_LIBRARIES ndbNDBT ndbclient_static perconaserverclient
+  LINK_LIBRARIES ndbNDBT ndbclient_static mysqlclient
   )
 
 IF(MY_COMPILER_IS_CLANG)
--- storage/ndb/tools/CMakeLists.txt
+++ storage/ndb/tools/CMakeLists.txt
@@ -230,10 +230,10 @@ IF(NOT WIN32 AND NOT SOLARIS)
       COMPONENT ClusterTools
       )
     IF(NCURSES_HAS_STDSCR)
-      TARGET_LINK_LIBRARIES(ndb_top perconaserverclient ${NCURSESW_LIB})
+      TARGET_LINK_LIBRARIES(ndb_top mysqlclient ${NCURSESW_LIB})
     ELSE()
       TARGET_LINK_LIBRARIES(ndb_top
-        perconaserverclient
+        mysqlclient
         ${NCURSESW_LIB}
         ${NCURSES_TINFO_LIB})
     ENDIF()
--- testclients/CMakeLists.txt
+++ testclients/CMakeLists.txt
@@ -24,9 +24,9 @@
 ADD_WSHADOW_WARNING()

 MYSQL_ADD_EXECUTABLE(mysql_client_test mysql_client_test.cc COMPONENT Test)
-TARGET_LINK_LIBRARIES(mysql_client_test perconaserverclient extra::rapidjson)
+TARGET_LINK_LIBRARIES(mysql_client_test mysqlclient extra::rapidjson)

 IF(WITH_UNIT_TESTS)
   MYSQL_ADD_EXECUTABLE(bug25714 bug25714.cc SKIP_INSTALL EXCLUDE_FROM_PGO)
-  TARGET_LINK_LIBRARIES(bug25714 perconaserverclient)
+  TARGET_LINK_LIBRARIES(bug25714 mysqlclient)
 ENDIF()
--- unittest/gunit/group_replication/CMakeLists.txt
+++ unittest/gunit/group_replication/CMakeLists.txt
@@ -49,7 +49,7 @@
     gr_unit_test_resource
     gunit_large
     server_unittest_library
-    perconaserverclient
+    mysqlclient
     mysqlgcs
     mysql_gtid
     )
--- unittest/gunit/xplugin/xcl/CMakeLists.txt
+++ unittest/gunit/xplugin/xcl/CMakeLists.txt
@@ -70,7 +70,7 @@
   harness_net_ts
   ext::libprotobuf
   ext::lz4
-  perconaserverclient
+  mysqlclient
   gtest
   gmock
 )
--- unittest/gunit/xplugin/xpl/CMakeLists.txt
+++ unittest/gunit/xplugin/xpl/CMakeLists.txt
@@ -115,7 +115,7 @@ TARGET_LINK_LIBRARIES(${XPL_UNIT_TESTS}
   ${GCOV_LDFLAGS}
   ${MYSQLX_CLIENT_LIB}
   mysqlx
-  perconaserverclient
+  mysqlclient
   gtest
   gmock
 )
#--- storage/tokudb/PerconaFT/tools/CMakeLists.txt       2018-08-29 12:23:52.000000000 -0400
#+++ storage/tokudb/PerconaFT/tools/CMakeLists.txt       2018-08-29 12:41:38.711986474 -0400
#@@ -20,7 +20,7 @@
#         target_link_libraries(${tool} sql binlog rpl master slave)
#       endif ()
#     else ()
#-      target_link_libraries(${tool} perconaserverclient)
#+      target_link_libraries(${tool} mysqlclient)
#     endif ()
#   endif ()

--- libmysql/authentication_kerberos/CMakeLists.txt     2021-09-28 04:49:14.000000000 +0000
+++ libmysql/authentication_kerberos/CMakeLists.txt     2021-09-28 09:24:09.917822028 +0000
@@ -65,7 +65,7 @@
   LIST(APPEND UTILITY "sspi_utility.cc")
 ENDIF()

-SET(MY_SQL "perconaserverclient")
+SET(MY_SQL "mysqlclient")

 DISABLE_MISSING_PROFILE_WARNING()

--- unittest/gunit/changestreams/CMakeLists.txt         2022-03-31 15:09:27.559009464 +0000
+++ unittest/gunit/changestreams/CMakeLists.txt         2022-03-31 15:10:26.404854666 +0000
@@ -36,7 +36,7 @@
 SET(TESTS_LINK_LIBRARIES
   changestreams_standalone_static
   mysql_binlog_event_standalone
-  perconaserverclient
+  mysqlclient
   gunit_small)

 IF (WITH_CS_PROTOBUF)
--- libchangestreams/src/tests/CMakeLists.txt   2023-10-21 17:41:00.756660313 +0300
+++ libchangestreams/src/tests/CMakeLists.txt   2023-10-21 17:41:14.344653893 +0300
@@ -38,7 +38,7 @@
 SET(TEST_LINK_LIBRARIES
   changestreams_standalone_static
   mysql_binlog_event_standalone
-  perconaserverclient
+  mysqlclient
   ${GTEST_LIBRARIES})

 # If we are using the protobuf extensions
--- libmysql/fido_client/common/CMakeLists.txt	2024-01-06 19:59:00.366685501 +0000
+++ libmysql/fido_client/common/CMakeLists.txt	2024-01-06 19:59:05.366685501 +0000
@@ -45,6 +45,6 @@
   fido_client_common
   ${FIDO_COMMON_SOURCES}
   LINK_LIBRARIES
-  perconaserverclient
+  mysqlclient
   ext::fido
 )
