--- client/base/CMakeLists.txt	2018-08-18 03:29:34.986169793 -0400
+++ client/base/CMakeLists.txt	2018-08-18 03:29:34.991586542 -0400
@@ -50,7 +50,7 @@
   ../../sql-common/sql_string.cc
 )
 
-TARGET_LINK_LIBRARIES(client_base perconaserverclient)
+TARGET_LINK_LIBRARIES(client_base mysqlclient)
 
 # Do not build library unless it is needed by some other target.
 SET_PROPERTY(TARGET client_base PROPERTY EXCLUDE_FROM_ALL TRUE)
--- client/CMakeLists.txt	2018-08-18 03:29:59.708213642 -0400
+++ client/CMakeLists.txt	2018-08-18 03:29:59.712547041 -0400
@@ -36,14 +36,14 @@
 INCLUDE(${MYSQL_CMAKE_SCRIPT_DIR}/compile_flags.cmake)
 
 MYSQL_ADD_EXECUTABLE(mysql completion_hash.cc mysql.cc readline.cc ../sql-common/sql_string.cc pattern_matcher.cc)
-TARGET_LINK_LIBRARIES(mysql perconaserverclient)
+TARGET_LINK_LIBRARIES(mysql mysqlclient)
 IF(UNIX)
   TARGET_LINK_LIBRARIES(mysql ${MY_READLINE_LIBRARY})
 ENDIF(UNIX)
 
 IF(NOT WITHOUT_SERVER)
   MYSQL_ADD_EXECUTABLE(mysql_upgrade upgrade/program.cc)
-  TARGET_LINK_LIBRARIES(mysql_upgrade perconaserverclient client_base mysqlcheck_core)
+  TARGET_LINK_LIBRARIES(mysql_upgrade mysqlclient client_base mysqlcheck_core)
   ADD_DEPENDENCIES(mysql_upgrade GenFixPrivs GenSysSchema GenBootstrapPriv)
 ENDIF()
 
@@ -52,24 +52,24 @@
 )
 
 MYSQL_ADD_EXECUTABLE(mysqltest mysqltest.cc ${MYSQLTEST_SRC} COMPONENT Test)
-TARGET_LINK_LIBRARIES(mysqltest perconaserverclient regex)
+TARGET_LINK_LIBRARIES(mysqltest mysqlclient regex)
 ADD_DEPENDENCIES(mysqltest GenError GenClientError)
 
 ADD_CONVENIENCE_LIBRARY(mysqlcheck_core check/mysqlcheck_core.cc)
-TARGET_LINK_LIBRARIES(mysqlcheck_core perconaserverclient)
+TARGET_LINK_LIBRARIES(mysqlcheck_core mysqlclient)
 
 MYSQL_ADD_EXECUTABLE(mysqlcheck check/mysqlcheck.cc)
 TARGET_LINK_LIBRARIES(mysqlcheck mysqlcheck_core)
 
 MYSQL_ADD_EXECUTABLE(mysqldump mysqldump.cc)
-TARGET_LINK_LIBRARIES(mysqldump perconaserverclient)
+TARGET_LINK_LIBRARIES(mysqldump mysqlclient)
 
 MYSQL_ADD_EXECUTABLE(mysqlimport mysqlimport.cc)
 SET_SOURCE_FILES_PROPERTIES(mysqlimport.cc PROPERTIES COMPILE_FLAGS "-DTHREADS")
-TARGET_LINK_LIBRARIES(mysqlimport perconaserverclient)
+TARGET_LINK_LIBRARIES(mysqlimport mysqlclient)
 
 MYSQL_ADD_EXECUTABLE(mysqlshow mysqlshow.cc)
-TARGET_LINK_LIBRARIES(mysqlshow perconaserverclient)
+TARGET_LINK_LIBRARIES(mysqlshow mysqlclient)
 
 MYSQL_ADD_EXECUTABLE(mysqlbinlog mysqlbinlog.cc)
 SET(MYSQLBINLOG_LIB_SOURCES
@@ -101,20 +101,20 @@
 )
 
 TARGET_LINK_LIBRARIES(mysqlbinlog mysqlbinlog_lib)
-TARGET_LINK_LIBRARIES(mysqlbinlog perconaserverclient binlogevents_static)
+TARGET_LINK_LIBRARIES(mysqlbinlog mysqlclient binlogevents_static)
 
 MYSQL_ADD_EXECUTABLE(mysqladmin mysqladmin.cc)
-TARGET_LINK_LIBRARIES(mysqladmin perconaserverclient)
+TARGET_LINK_LIBRARIES(mysqladmin mysqlclient)
 
 MYSQL_ADD_EXECUTABLE(mysqlslap mysqlslap.cc)
 SET_SOURCE_FILES_PROPERTIES(mysqlslap.cc PROPERTIES COMPILE_FLAGS "-DTHREADS")
-TARGET_LINK_LIBRARIES(mysqlslap perconaserverclient)
+TARGET_LINK_LIBRARIES(mysqlslap mysqlclient)
 
 MYSQL_ADD_EXECUTABLE(mysql_config_editor mysql_config_editor.cc)
-TARGET_LINK_LIBRARIES(mysql_config_editor perconaserverclient)
+TARGET_LINK_LIBRARIES(mysql_config_editor mysqlclient)
 
 MYSQL_ADD_EXECUTABLE(mysql_secure_installation mysql_secure_installation.cc)
-TARGET_LINK_LIBRARIES(mysql_secure_installation perconaserverclient)
+TARGET_LINK_LIBRARIES(mysql_secure_installation mysqlclient)
 
 MYSQL_ADD_EXECUTABLE(mysql_ssl_rsa_setup mysql_ssl_rsa_setup.cc path.cc logger.cc)
 TARGET_LINK_LIBRARIES(mysql_ssl_rsa_setup mysys mysys_ssl)
--- cmake/libutils.cmake	2018-08-18 03:30:07.106409849 -0400
+++ cmake/libutils.cmake	2018-08-18 03:30:07.110743248 -0400
@@ -33,8 +33,8 @@
 # reduce excessive recompiles.
 
 # Except for convenience libraries, this file provides macros to merge static 
-# libraries (we need it for perconaserverclient) and to create shared library out of 
-# convenience libraries(again, for perconaserverclient)
+# libraries (we need it for mysqlclient) and to create shared library out of 
+# convenience libraries(again, for mysqlclient)
 
 
 GET_FILENAME_COMPONENT(MYSQL_CMAKE_SCRIPT_DIR ${CMAKE_CURRENT_LIST_FILE} PATH)
--- include/mysql.h	2018-08-18 03:30:11.721480261 -0400
+++ include/mysql.h	2018-08-18 03:30:11.727980360 -0400
@@ -28,7 +28,7 @@
 /**
   @file include/mysql.h
   This file defines the client API to MySQL and also the ABI of the
-  dynamically linked libperconaserverclient.
+  dynamically linked libmysqlclient.
 
   The ABI should never be changed in a released product of MySQL,
   thus you need to take great care when changing the file. In case
@@ -399,12 +399,12 @@
 
 /*
   mysql_server_init/end need to be called when using libmysqld or
-  libperconaserverclient (exactly, mysql_server_init() is called by
+  libmysqlclient (exactly, mysql_server_init() is called by
   mysql_init() so you don't need to call it explicitely; but you need to call
   mysql_server_end() to free memory). The names are a bit misleading
   (mysql_SERVER* to be used when using libmysqlCLIENT). So we add more general
   names which suit well whether you're using libmysqld or
-  libperconaserverclient. We intend to promote these aliases over the
+  libmysqlclient. We intend to promote these aliases over the
   mysql_server* ones.
 */
 #define mysql_library_init mysql_server_init
--- libmysql/CMakeLists.txt	2018-08-18 03:30:19.520515918 -0400
+++ libmysql/CMakeLists.txt	2018-08-18 03:30:19.525932667 -0400
@@ -238,13 +238,13 @@
 MESSAGE(STATUS "Creating LDAP authentication SASL client library.")
 ADD_SUBDIRECTORY(authentication_ldap)
   
-# Merge several convenience libraries into one big perconaserverclient
-MERGE_CONVENIENCE_LIBRARIES(perconaserverclient ${LIBS_TO_MERGE} COMPONENT Development)
-TARGET_LINK_LIBRARIES(perconaserverclient ${LIBS_TO_LINK})
+# Merge several convenience libraries into one big mysqlclient
+MERGE_CONVENIENCE_LIBRARIES(mysqlclient ${LIBS_TO_MERGE} COMPONENT Development)
+TARGET_LINK_LIBRARIES(mysqlclient ${LIBS_TO_LINK})
 
 # Visual Studio users need debug  static library for debug projects
 IF(MSVC)
-  INSTALL_DEBUG_TARGET(perconaserverclient DESTINATION ${INSTALL_LIBDIR}/debug)
+  INSTALL_DEBUG_TARGET(mysqlclient DESTINATION ${INSTALL_LIBDIR}/debug)
 
   INSTALL_DEBUG_TARGET(auth_win_client DESTINATION ${INSTALL_LIBDIR}/debug)
   INSTALL_DEBUG_TARGET(clientlib DESTINATION ${INSTALL_LIBDIR}/debug)
@@ -276,7 +276,7 @@
 ENDIF()
 
 IF(NOT DISABLE_SHARED)
-  # Merge several convenience libraries into one big perconaserverclient
+  # Merge several convenience libraries into one big mysqlclient
   # and link them together into shared library.
   MERGE_LIBRARIES_SHARED(libmysql ${LIBS_TO_MERGE}
     EXPORTS ${CLIENT_API_FUNCTIONS} ${CLIENT_API_FUNCTIONS_UNDOCUMENTED}
@@ -299,9 +299,9 @@
       SET(OS_SHARED_LIB_VERSION
         "${SHARED_LIB_MAJOR_VERSION}.${SHARED_LIB_MINOR_VERSION}.${SHARED_LIB_PATCH_VERSION}")
     ENDIF()
-    # Name of shared library is perconaserverclient on Unix
+    # Name of shared library is mysqlclient on Unix
     SET_TARGET_PROPERTIES(libmysql PROPERTIES 
-      OUTPUT_NAME perconaserverclient
+      OUTPUT_NAME mysqlclient
       VERSION "${OS_SHARED_LIB_VERSION}" 
       SOVERSION "${SHARED_LIB_MAJOR_VERSION}")
     CONFIGURE_FILE(libmysql.ver.in ${CMAKE_CURRENT_BINARY_DIR}/libmysql.ver)
@@ -322,8 +322,8 @@
         PROPERTIES LINK_FLAGS "${libmysql_link_flags}")
     ENDIF() 
     # clean direct output needs to be set several targets have the same name
-    #(perconaserverclient in this case)
-    SET_TARGET_PROPERTIES(perconaserverclient PROPERTIES CLEAN_DIRECT_OUTPUT 1)
+    #(mysqlclient in this case)
+    SET_TARGET_PROPERTIES(mysqlclient PROPERTIES CLEAN_DIRECT_OUTPUT 1)
     SET_TARGET_PROPERTIES(libmysql PROPERTIES CLEAN_DIRECT_OUTPUT 1)
 
   ENDIF()
--- libmysql/libmysql.ver.in	2018-08-18 03:31:20.419945051 -0400
+++ libmysql/libmysql.ver.in	2018-08-18 03:31:20.424278451 -0400
@@ -21,5 +21,5 @@
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
 
 
-libperconaserverclient_@SHARED_LIB_MAJOR_VERSION@.@SHARED_LIB_MINOR_VERSION@
+libmysqlclient_@SHARED_LIB_MAJOR_VERSION@.@SHARED_LIB_MINOR_VERSION@
 { global: ${CLIENT_API_FUNCTIONS};${CLIENT_API_FUNCTIONS_UNDOCUMENTED}; local: *; };
--- scripts/CMakeLists.txt	2018-08-18 03:31:39.003728628 -0400
+++ scripts/CMakeLists.txt	2018-08-18 03:31:39.009145377 -0400
@@ -365,7 +365,7 @@
 ENDIF()
 
 # Use cmake variables to inspect dependencies for 
-# perconaserverclient library (add -l stuff)
+# mysqlclient library (add -l stuff)
 SET(CLIENT_LIBS "")
 SET(LIBS "")
 
@@ -398,11 +398,11 @@
   STRING(REGEX REPLACE "[ ]+$" "" ${var} "${${var}}")
 ENDMACRO()
 
-EXTRACT_LINK_LIBRARIES(perconaserverclient CLIENT_LIBS)
+EXTRACT_LINK_LIBRARIES(mysqlclient CLIENT_LIBS)
 
 IF(MSVC)
-  GET_TARGET_PROPERTY(LIBMYSQL_OS_SHLIB_VERSION perconaserverclient VERSION)
-  GET_TARGET_PROPERTY(LIBMYSQL_OS_OUTPUT_NAME perconaserverclient OUTPUT_NAME)
+  GET_TARGET_PROPERTY(LIBMYSQL_OS_SHLIB_VERSION mysqlclient VERSION)
+  GET_TARGET_PROPERTY(LIBMYSQL_OS_OUTPUT_NAME mysqlclient OUTPUT_NAME)
 ELSE()
   GET_TARGET_PROPERTY(LIBMYSQL_OS_SHLIB_VERSION libmysql VERSION)
   GET_TARGET_PROPERTY(LIBMYSQL_OS_OUTPUT_NAME libmysql OUTPUT_NAME)
--- storage/ndb/ndbapi-examples/mgmapi_logevent/Makefile	2018-08-18 03:31:53.867288716 -0400
+++ storage/ndb/ndbapi-examples/mgmapi_logevent/Makefile	2018-08-18 03:31:53.870538766 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(OBJS): $(SRCS)
 	$(CXX) $(CFLAGS) -I$(TOP_SRCDIR)/include -I$(INCLUDE_DIR) -I$(INCLUDE_DIR)/mgmapi  -I$(INCLUDE_DIR)/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/mgmapi_logevent2/Makefile	2018-08-18 03:32:01.879744295 -0400
+++ storage/ndb/ndbapi-examples/mgmapi_logevent2/Makefile	2018-08-18 03:32:01.884077695 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(OBJS): $(SRCS)
 	$(CXX) $(CFLAGS) -I$(TOP_SRCDIR)/include -I$(INCLUDE_DIR) -I$(INCLUDE_DIR)/mgmapi  -I$(INCLUDE_DIR)/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_async/Makefile	2018-08-18 03:32:11.933231001 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_async/Makefile	2018-08-18 03:32:11.936481050 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(TARGET).o: $(SRCS)
 	$(CXX) $(CFLAGS) -I$(INCLUDE_DIR)/include -I$(INCLUDE_DIR)/storage/ndb/include -I$(INCLUDE_DIR)/storage/ndb/include/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_async1/Makefile	2018-08-18 03:32:17.497315889 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_async1/Makefile	2018-08-18 03:32:17.501649289 -0400
@@ -35,7 +35,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(TARGET).o: $(SRCS)
 	$(CXX) $(CFLAGS) -I$(TOP_SRCDIR)/include -I$(INCLUDE_DIR) -I$(INCLUDE_DIR)/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_blob/Makefile	2018-08-18 03:32:25.159849464 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_blob/Makefile	2018-08-18 03:32:25.163099513 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(TARGET).o: $(SRCS)
 	$(CXX) $(CFLAGS) -I$(INCLUDE_DIR)/include -I$(INCLUDE_DIR)/storage/ndb/include -I$(INCLUDE_DIR)/storage/ndb/include/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_blob_ndbrecord/Makefile	2018-08-18 03:32:34.693328250 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_blob_ndbrecord/Makefile	2018-08-18 03:32:34.697661649 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(OBJS): $(SRCS)
 	$(CXX) $(CFLAGS) -I$(INCLUDE_DIR)/include -I$(INCLUDE_DIR)/storage/ndb/include -I$(INCLUDE_DIR)/storage/ndb/include/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_event/Makefile	2018-08-18 03:32:39.731988459 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_event/Makefile	2018-08-18 03:32:39.736321858 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(TARGET).o: $(SRCS) Makefile
 	$(CXX) $(CFLAGS) $(DEBUG) -I$(INCLUDE_DIR) -I$(INCLUDE_DIR)/ndbapi -I$(TOP_SRCDIR)/include $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_recattr_vs_record/Makefile	2018-08-18 03:32:51.553502151 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_recattr_vs_record/Makefile	2018-08-18 03:32:51.558918901 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(OBJS): $(SRCS)
 	$(CXX) $(CFLAGS) -I$(INCLUDE_DIR)/include -I$(INCLUDE_DIR)/storage/ndb/include -I$(INCLUDE_DIR)/storage/ndb/include/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_retries/Makefile	2018-08-18 03:33:03.477934084 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_retries/Makefile	2018-08-18 03:33:03.483350833 -0400
@@ -35,7 +35,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(TARGET).o: $(SRCS)
 	$(CXX) $(CFLAGS)  -I$(TOP_SRCDIR)/include -I$(INCLUDE_DIR) -I$(INCLUDE_DIR)/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_scan/Makefile	2018-08-18 03:33:11.319220386 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_scan/Makefile	2018-08-18 03:33:11.322470436 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(TARGET).o: $(SRCS)
 	$(CXX) $(CFLAGS) -I$(INCLUDE_DIR)/include -I$(INCLUDE_DIR)/storage/ndb/include -I$(INCLUDE_DIR)/storage/ndb/include/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_simple/Makefile	2018-08-18 03:33:20.103021067 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_simple/Makefile	2018-08-18 03:33:20.107354466 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(TARGET).o: $(SRCS)
 	$(CXX) $(CFLAGS) -I$(TOP_SRCDIR)/include -I$(INCLUDE_DIR) -I$(INCLUDE_DIR)/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_simple_dual/Makefile	2018-08-18 03:33:28.880321646 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_simple_dual/Makefile	2018-08-18 03:33:28.884655046 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(OBJS): $(SRCS)
 	$(CXX) $(CFLAGS) -I$(TOP_SRCDIR)/include -I$(INCLUDE_DIR) -I$(INCLUDE_DIR)/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_simple_index/Makefile	2018-08-18 03:33:37.747540266 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_simple_index/Makefile	2018-08-18 03:33:37.752957016 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(OBJS): $(SRCS)
 	$(CXX) $(CFLAGS) -I$(INCLUDE_DIR)/include -I$(INCLUDE_DIR)/storage/ndb/include -I$(INCLUDE_DIR)/storage/ndb/include/ndbapi $(SRCS)
--- storage/ndb/ndbapi-examples/ndbapi_s_i_ndbrecord/Makefile	2018-08-18 03:33:46.284337179 -0400
+++ storage/ndb/ndbapi-examples/ndbapi_s_i_ndbrecord/Makefile	2018-08-18 03:33:46.289753929 -0400
@@ -36,7 +36,7 @@
 SYS_LIB = 
 
 $(TARGET): $(OBJS)
-	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lperconaserverclient_r $(SYS_LIB) -o $(TARGET)
+	$(CXX) $(CXXFLAGS) $(LFLAGS) $(LIB_DIR) $(OBJS) -lndbclient -lmysqlclient_r $(SYS_LIB) -o $(TARGET)
 
 $(OBJS): $(SRCS)
 	$(CXX) $(CFLAGS) -I$(INCLUDE_DIR)/include -I$(INCLUDE_DIR)/storage/ndb/include -I$(INCLUDE_DIR)/storage/ndb/include/ndbapi $(SRCS)
--- storage/ndb/test/run-test/CMakeLists.txt	2018-08-18 03:33:54.833050942 -0400
+++ storage/ndb/test/run-test/CMakeLists.txt	2018-08-18 03:33:54.837384341 -0400
@@ -39,7 +39,7 @@
 MYSQL_ADD_EXECUTABLE(atrt main.cpp setup.cpp files.cpp db.cpp command.cpp
   DESTINATION mysql-test/ndb
   )
-TARGET_LINK_LIBRARIES(atrt ndbNDBT ndbclient_static perconaserverclient)
+TARGET_LINK_LIBRARIES(atrt ndbNDBT ndbclient_static mysqlclient)
 
 FILE(GLOB testcase_files RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} "*-tests.txt")
 ADD_TEST(NAME check_testcase_files
--- testclients/CMakeLists.txt	2018-08-18 03:34:01.038478951 -0400
+++ testclients/CMakeLists.txt	2018-08-18 03:34:01.042812350 -0400
@@ -21,11 +21,11 @@
 # Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
 
 MYSQL_ADD_EXECUTABLE(mysql_client_test mysql_client_test.cc COMPONENT Test)
-TARGET_LINK_LIBRARIES(mysql_client_test perconaserverclient)
+TARGET_LINK_LIBRARIES(mysql_client_test mysqlclient)
 SET_TARGET_PROPERTIES(mysql_client_test PROPERTIES LINKER_LANGUAGE CXX)
 
 IF(WITH_UNIT_TESTS)
   MYSQL_ADD_EXECUTABLE(bug25714 bug25714.cc SKIP_INSTALL)
-  TARGET_LINK_LIBRARIES(bug25714 perconaserverclient)
+  TARGET_LINK_LIBRARIES(bug25714 mysqlclient)
   SET_TARGET_PROPERTIES(bug25714 PROPERTIES LINKER_LANGUAGE CXX)
 ENDIF()
--- plugin/percona-pam-for-mysql/CMakeLists.txt_new     2018-08-21 07:01:38.155232659 -0400
+++ plugin/percona-pam-for-mysql/CMakeLists.txt 2018-08-21 07:02:18.194400527 -0400
@@ -37,7 +37,7 @@
   MYSQL_ADD_PLUGIN(dialog
                    src/dialog.cc
                    ../../client/get_password.cc
-                   LINK_LIBRARIES perconaserverclient
+                   LINK_LIBRARIES mysqlclient
                    MODULE_ONLY)
 ENDIF(HAVE_PAM AND HAVE_GETPWNAM_R AND HAVE_GETGRGID_R AND HAVE_DLFCN_H)
 ENDIF(WITH_PAM)
--- unittest/gunit/group_replication/CMakeLists.txt     2018-08-22 01:55:48.775894920 -0400
+++ unittest/gunit/group_replication/CMakeLists.txt     2018-08-22 01:56:19.804229110 -0400
@@ -45,7 +45,7 @@
   TARGET_LINK_LIBRARIES(${test}-t gr_unit_test_resource gunit_small ${GCS_LIBRARY})
   # mysys_ssl/libmysys_ssl.a already includes SSL objects on Solaris
   IF(NOT CMAKE_SYSTEM_NAME MATCHES "SunOS")
-    TARGET_LINK_LIBRARIES(${test}-t perconaserverclient)
+    TARGET_LINK_LIBRARIES(${test}-t mysqlclient)
   ENDIF()
 ENDFOREACH()

--- unittest/gunit/xplugin/CMakeLists.txt	2018-08-22 02:34:45.000000000 -0400
+++ unittest/gunit/xplugin/CMakeLists.txt 	2018-08-22 02:55:10.561454824 -0400
@@ -97,7 +97,7 @@
   ${GCOV_LDFLAGS}
   ${MYSQLX_CLIENT_LIB}
   mysqlx
-  perconaserverclient
+  mysqlclient
   gtest
   gmock
 )
@@ -142,7 +142,7 @@
   ${MYSQLX_CLIENT_FULL_LIB}
   ${MYSQLX_PROTOCOL_FULL_LIB}
   ${PROTOBUF_LIBRARY}
-  perconaserverclient
+  mysqlclient
   gtest
   gmock
 )
--- plugin/x/tests/driver/CMakeLists.txt	2018-08-22 03:58:32.371339263 -0400
+++ plugin/x/tests/driver/CMakeLists.txt	2018-08-22 03:58:46.195045928 -0400
@@ -67,7 +67,7 @@
 TARGET_LINK_LIBRARIES(${XTEST_EXE_NAME}
   mysys
   strings
-  perconaserverclient
+  mysqlclient
   ${MYSQLX_CLIENT_FULL_LIB}
   ${MYSQLX_PROTOCOL_FULL_LIB}
   ${PROTOBUF_LIBRARY}
--- storage/tokudb/PerconaFT/tools/CMakeLists.txt       2018-08-29 12:23:52.000000000 -0400
+++ storage/tokudb/PerconaFT/tools/CMakeLists.txt       2018-08-29 12:41:38.711986474 -0400
@@ -20,7 +20,7 @@
         target_link_libraries(${tool} sql binlog rpl master slave)
       endif ()
     else ()
-      target_link_libraries(${tool} perconaserverclient)
+      target_link_libraries(${tool} mysqlclient)
     endif ()
   endif ()
 
--- router/CMakeLists.txt       2018-12-11 10:52:17.122381407 +0200
+++ router/CMakeLists.txt       2018-12-11 10:52:17.122381407 +0200
@@ -95,7 +95,7 @@
 INCLUDE(cmake/Plugin.cmake)
 
 # Set the proper paths for the mysqlclient library
-SET(MySQL_CLIENT_LIB perconaserverclient)
+SET(MySQL_CLIENT_LIB mysqlclient)
 SET(MySQL_VERSION_HEADER ${CMAKE_BINARY_DIR}/include/mysql_version.h)
 SET(MySQL_INCLUDE_DIRS ${CMAKE_SOURCE_DIR}/include)
 LIST(APPEND MySQL_INCLUDE_DIRS ${CMAKE_SOURCE_DIR}/libbinlogevents/export)

